### PR TITLE
Add spans for index DB reads [LPF-270]

### PR DESCRIPTION
**Pull request by curiousleo-da**
Emit OpenTelemetry spans and events for index DB reads.

Screenshot_2020-12-10 Jaeger UI(2)

Screenshot_2020-12-10 Jaeger UI

Searching for correlated calls via a given `daml.offset`:

Screenshot_2020-12-10 Jaeger UI(1)

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

### CODE REVIEW

**Comment by curiousleo-da** [comment]
@lucb-da, could you take a look?

This is a very straightforward, abstraction-less way to get the traces. I consciously avoided creating a tracing library inside `daml` for this PR.

I'm aware that you are working on such a tracing library for this repo.

What are your thoughts on the best way forward here -- do you think that this PR (with its lack of abstraction) can be merged and then made to use the tracing library once it has landed, or does that sound like a lot of additional effort?

**Comment by curiousleo-da** [comment]
@lucb-da, @gerolf-da PTAL.

**Review by lucb-da — APPROVED** [review]
Looks good in general.
The resulting traces are great. I think a future improvement would be to add information about who is requesting the transactions.
I don't see any problem on merging this code, and then include it in the future tracing support.

File: ledger/metrics/src/main/scala/com/daml/metrics/Event.scala [hunk 1]
```diff
@@ -14,7 +14,9 @@ final case class Event(name: String, attributeMap: Map[String, String])
 
   override lazy val getAttributes: Attributes = {
     val attributes = Attributes.newBuilder()
-    attributeMap.foreach { case (k, v) => attributes.setAttribute(k, v) }
+    attributeMap
+      .filter { case (_, v) => !v.isEmpty }
+      .foreach { case (k, v) => attributes.setAttribute(k, v) }
```
[the remark this answers is not in the document]
Comment (curiousleo-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/Event.scala:19]: I want to filter out some elements, so using `.filter` seems appropriate :)

This pertains to stylistic conventions in the repo that I'm not well aware of (and perhaps performance considerations?). Your comment suggests that slightly larger function bodies with fewer chained collection calls (just `.foreach`) are preferred over minimal function bodies with more chained collection calls (`.filter` + `.foreach`).

I'll move the condition inside the `.foreach`.

File: ledger/metrics/src/main/scala/com/daml/metrics/package.scala [hunk 2]
```diff
@@ -3,11 +3,15 @@
...
 
-import com.codahale.metrics.{Gauge, MetricRegistry}
 import com.codahale.metrics.MetricRegistry.MetricSupplier
+import com.codahale.metrics.{Gauge, MetricRegistry}
+import io.opentelemetry.OpenTelemetry
+import io.opentelemetry.trace.Tracer
 
 package object metrics {
 
+  val OpenTelemetryTracer: Tracer = OpenTelemetry.getTracer("ledger")
```
[the remark this answers is not in the document]
Comment (curiousleo-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/package.scala]: :+1: I've changed both.

**Review by gerolf-da — APPROVED** [review]
Thanks, @curiousleo-da!

**Pull request merged**